### ISSUE TITLE
Question about OSSL_PROVIDER_set_default_search_path and OSSL_PROVIDER_available

### ISSUE THREAD

**Issue by keyslapper** [openssl/openssl#18212]
Hey everyone,

I'm asking this here because it seems like the behavior I'm seeing doesn't match the documentation I'm reading.  Perhaps I'm not understanding something correctly though - wouldn't be the first time.

My code uses OpenSSL, on several platforms, and the FIPS provider on a subset of those platforms.  For those platforms, we need to make sure the fips provider we supply with the software is the one being loaded, which is easy enough in production - we just build with a different install destination that matches that of our software.

However, given other resources used by the software, we cannot use this in development situations.  We have several devs working on the same hosts, using somewhat isolated configurations.  When we update our OpenSSL build for a service patch, we need to be able to test it without disrupting other work going on.  So there is a reason for our non-standard approach here.  So programmatically, we need to be able to load both configs and providers from non-standard locations, even different from the OpenSSL build configuration.

Here's roughly what I'm doing:

```
if (CONF_modules_load_file("/tmp/testbed/config/openssl.cnf", NULL, 0) != 1)
  {
    LOG("Config file load failed");
    return ERROR_CONFIG_FAILURE;
  }
...
osslLibCtx = OSSL_LIB_CTX_get0_global_default();
OSSL_PROVIDER_set_default_search_path(osslLibCtx, "/tmp/testbed/ossl-modules/");
if (OSSL_PROVIDER_available(osslLibCtx, "fips") != 1)
  {
    LOG("FIPS provider not available in default search path");
    return ERROR_NO_FIPS;
  }
fProvider = OSSL_PROVIDER_load(osslLibCtx, "fips);
if (fProvider != NULL)
  {
    bProvider = OSSL_PROVIDER_load(osslLibCtx, "base);
  }
if ((fProvider == NULL) || (bProvider == NULL))
  {
    LOG("Provider load failure");
    return ERROR_PROVIDER;
  }
if (EVP_default_properties_enable_fips(osslLibCtx, 1) == 0)
  {
    LOG("FIPS init error");
    return ERROR_FIPS_INIT;
  }
if (EVP_default_properties_is_fips_enabled(osslLibCtx) == 1)
  {
    LOG("FIPS has been enabled");
  }
else
  {
    LOG("FIPS has not been enabled");
    return ERROR_FIPS_INIT;
  }

```

I get that some of this might seem redundant, but I've been working on it awhile and added a LOT of verification and logging to try to narrow down where I've been going wrong.

It took some time to realize what the configuration issues were, but I eventually just added the [fips_sect] config form fipsmodule.cnf to my openssl.cnf - again, in a non-standard location.  Because I'm explicitly loading the providers, I've removed all the activation directives from my provider sections.

The problem I can't seem to get around is the **OSSL_PROVIDER_available()** call.  The OSSL_PROVIDER_set_default_search_path() call does make OpenSSL look for the right fips.so provider module, but OSSL_PROVIDER_available does **NOT** seem to use that path - that call ALWAYS fails.  If I ignore the results of that check and try to load fips.so anyway, it works fine.

Am I setting something up incorrectly here?  Anything you can tell me would be helpful.

Thank you
Lou

**Comment by mattcaswell** [comment]
> if (CONF_modules_load_file("/tmp/testbed/config/openssl.cnf", NULL, 0) != 1)

A (better?) way to do this might be `OSSL_LIB_CTX_load_config`:

https://www.openssl.org/docs/man3.0/man3/OSSL_LIB_CTX_load_config.html

Or you can set the `OPENSSL_CONF` environment variable to point to the correct location - and that will override the default location and automatically load the config file from there instead (into the default libctx).

> osslLibCtx = OSSL_LIB_CTX_get0_global_default();

This isn't necessary. You can just use `NULL` to refer to the global default libctx.

> OSSL_PROVIDER_set_default_search_path(osslLibCtx, "/tmp/testbed/ossl-modules/");

An alternative way to do this is to set the `OPENSSL_MODULES` environment variable before starting your process.

> OSSL_PROVIDER_available does NOT seem to use that path - that call ALWAYS fails

`OSSL_PROVIDER_available` does not tell you whether the provider is in the search path. It tells you whether the provider has been loaded and is therefore available for algorithm fetches. If you call this *after* you've loaded the provider I would expect it to return true.

**Comment by keyslapper** [comment]
Thank you for the response Matt,

I think I'll go back to using NULL for the default context instead of grabbing it explicitly - I was doing that to rule out anything else I might have misunderstood in the first place anyway.

And I think I like the alternative of using OSSL_LIB_CTX_load_config better, thank you for the suggestion.

I've tried the OPENSSL_MODULES environment variable, but something was interfering with that - I think I read something about permissions.  Some of the processes that use the library we access OpenSSL through run as root, others do not.  Not sure if that's what the hangup was, but rather than work through a potentially convoluted set of config rules in our environments I opted for the direct route - and that path will end up being configurable under certain circumstances, but unmodifiable in production.

I think I understand what my misunderstanding was, with OSSL_PROVIDER_available, now, thank you for explaining!

Cheers!
Lou
